zepid/causal/gformula/TimeFixed.py

```diff
@@ -221,3 +221,90 @@ def fit(self, treatment):
         else:  # weighted marginal estimate
             self.marginal_outcome = np.average(g[self.outcome], weights=self.gf[self._weights])
         self.predicted_df = g
+
+    def fit_stochastic(self, p, conditional=None, samples=100, seed=None):
+        """Fits the g-formula for a stochastic intervention. As currently implemented, 'p' percent of the population is
+        randomly treated. This process is repeated 'n' times and the mean is the marginal stochastic outcome
+
+        Parameters
+        ----------
+        p: float, list
+            Percent of the population to randomly treat
+        conditional:
+            List of selection #TODO write this better
+        samples: int, optional
+            Number of resamples to calculate the marginal outcome. Default is 100
+        seed: int, optional
+            Seed for the random process selection
+        """
+        if self._outcome_model is None:
+            raise ValueError('Before the g-formula can be calculated, the outcome model must be specified')
+        if self.exposure_type != 'binary':
+            raise ValueError('Only binary exposures are currently available for the stochastic treatment g-formula')
+        if conditional is not None:
+            if len(p) != len(conditional):
+                raise ValueError("'p' and 'conditional' must be the same length")
+        else:
+            if type(p) != float:
+                raise ValueError("Specified percent must be a float when 'conditional' is not specified")
+
+        if seed is None:
+            pass
+        else:
+            np.random.seed(seed)
+
+        if conditional is not None:
+            self._check_conditional(conditional)
+
+        marginals = []
+        for s in range(samples):
+            g = self.gf.reset_index(drop=True).copy()
+
+            # Selecting out observations
+            if conditional is None:  # Unconditional probabilities
+                if self._weights is None:
+                    treated = np.random.choice(g.index, size=int(p*g.shape[0]), replace=False)
+                else:  # Weighting observations for random treatment
+                    treated = np.random.choice(g.index, size=int(p*g.shape[0]), replace=False,
+                                               p=g[self._weights]/np.sum(g[self._weights]))
+            else:  # Conditional probabilities based on eval()
+                # TODO Checking conditionals are mutually exclusive
+                if self._weights is None:
```

Comment:
If I squint as these two pieces of code, they are about identical. I don't know if you have additional plans for these branches, but what about:
```python
treated = []
for c, prop in zip(conditional, p):
    gs = g.loc[eval(c)].copy()
    p = gs[self._weights] / np.sum(gs[self._weights]) if self._weights else None
    tr = np.random.choice(gs.index, size=int(prop*gs.shape[0]), replace=False, p=p)
    treated.extend(tr)
```

Reply:
Ditto with the conditional above too

Reply:
Ahhh good call. I remember now that is how ``lifelines`` is set up. It is much more condensed and avoids my needless repitition